From aac81a3b605f20e19c49117869aca69e47989cfd Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex@uq.edu.au>
Date: Fri, 9 Aug 2019 14:06:58 +1000
Subject: [PATCH] joyent/node-mahi#3 filtering of roles to ones from the owning
 account should be done in libmahi, not in muskie/marlin

---
 lib/client.js | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/lib/client.js b/lib/client.js
index 363df52..fff213c 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -536,8 +536,21 @@ MahiClient.prototype.authorize = function authorize(opts) {
              * Otherwise ignore this role membership -- admin roles can't have
              * any attached policy.
              */
+            continue;
+        }
+
+        /*
+         * We do this here and not before the ADMIN_ROLE_NAME check because of
+         * the isOperator inheritance of sub-users on an operator account.
+         *
+         * In all other cases, we never want to authorize an action based on
+         * a role that belongs to a different account.
+         */
+        if (owner.account.uuid !== role.account) {
+            continue;
+        }
 
-        } else if (resourceTags.indexOf(activeRoles[i]) >= 0) {
+        if (resourceTags.indexOf(activeRoles[i]) >= 0) {
             /* Ordinary non-admin roles are processed below. */
             matchingRoles.push(activeRoles[i]);
             role.rules.forEach(function (rule) {
-- 
2.21.0

